### ISSUE TITLE
[BUG] RTSP Url stream presumed incompatibility in Maui.Toolkit Android Exoplayer component

### ISSUE THREAD

**Issue by Anioz** [CommunityToolkit/Maui#972]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Did you read the "Reporting a bug" section on Contributing file?

- [X] I have read the "Reporting a bug" section on Contributing file: https://github.com/CommunityToolkit/Maui/blob/main/CONTRIBUTING.md#reporting-a-bug

### Current Behavior

Crashes the app on the latest version at current time with following error:

"
[AndroidRuntime] FATAL EXCEPTION: ExoPlayer:Playback
[AndroidRuntime] Process: com.solid.HomeCon, PID: 18345
[AndroidRuntime] java.lang.IllegalArgumentException
[AndroidRuntime] 	at com.google.android.exoplayer2.util.Assertions.checkArgument(Assertions.java:39)
[AndroidRuntime] 	at com.google.android.exoplayer2.source.rtsp.RtspMediaTrack.generatePayloadFormat(RtspMediaTrack.java:260)
[AndroidRuntime] 	at com.google.android.exoplayer2.source.rtsp.RtspMediaTrack.<init>(RtspMediaTrack.java:161)
[AndroidRuntime] 	at com.google.android.exoplayer2.source.rtsp.RtspClient.buildTrackList(RtspClient.java:347)
[AndroidRuntime] 	at com.google.android.exoplayer2.source.rtsp.RtspClient.access$1700(RtspClient.java:75)
[AndroidRuntime] 	at com.google.android.exoplayer2.source.rtsp.RtspClient$MessageListener.onDescribeResponseReceived(RtspClient.java:690)
[AndroidRuntime] 	at com.google.android.exoplayer2.source.rtsp.RtspClient$MessageListener.handleRtspResponse(RtspClient.java:598)
[AndroidRuntime] 	at com.google.android.exoplayer2.source.rtsp.RtspClient$MessageListener.handleRtspMessage(RtspClient.java:507)
[AndroidRuntime] 	at com.google.android.exoplayer2.source.rtsp.RtspClient$MessageListener.lambda$onRtspMessageReceived$0$com-google-android-exoplayer2-source-rtsp-RtspClient$MessageListener(RtspClient.java:500)
[AndroidRuntime] 	at com.google.android.exoplayer2.source.rtsp.RtspClient$MessageListener$$ExternalSyntheticLambda0.run(Unknown Source:4)
[AndroidRuntime] 	at android.os.Handler.handleCallback(Handler.java:883)
[AndroidRuntime] 	at android.os.Handler.dispatchMessage(Handler.java:100)
[AndroidRuntime] 	at android.os.Looper.loop(Looper.java:237)
[AndroidRuntime] 	at android.os.HandlerThread.run(HandlerThread.java:67)
[m.solid.HomeCo] 0xebadde09 skipped times: 0
[libc] exiting due to SIG_DFL handler for signal 11"

Hikvision IP camera model HWI-D140H x 3 works great without any problems  (maui toolkit rtsp://user:pswd@ip:554/ch1/sub/live) 
Hikvision Doorbell camera model DS-KD8003-IME1 x1 crashes the app (maui toolkit rtsp://user:pswd@ip:554/ch1/sub/live works both in vlc and xamarin forms)
Foscam IP camera model R4M x 1 crashes the app (maui toolkit rtsp://user:pswd@ip:88/ch1/sub/live, works both in vlc and xamarin forms) 
Foscam IP camera model R4M-B x 1 crashes the app (maui toolkit rtsp://user:pswd@ip:88/ch1/sub/live, works both in vlc and xamarin forms)

Ive tested with other ports this seems NOT to be the cause of the problem since the hikvision Doorbell crashes with the same error and uses the same url as the hikvision ip camera's (ive also tested a video only stream on the doorbell and it fails with the same exception)

Please advise on what tests to additionally try and if this will be looked at in an upcoming release of maui toolkit mediaelement, since at current time theres no real viable alternative

### Expected Behavior

Plays any rtsp URL as long as given url format is a valid working one.

### Steps To Reproduce

1. Create new maui .net 7 framework solution
2. Install maui toolkit nuget and follow readme instructions
3. Add mediaelement xmlns and the element itself to one of the contentpages
4. Prove to yourself that the specific given model of the Foscam and/or Hikvision camera is up and running with any other validation method (for example vlc or a previous xamarin forms project)
5. Set the mediaelement source as the proven working rtsp url
6. Run your Android app (debug or release)
7. Check the log for the above fatal app crash

### Link to public reproduction project repository

https://1drv.ms/u/s!AqVY9RkzgdHIrRJpCfa_rIgyBW7A?e=8oxTBc

### Environment

```markdown
- .NET MAUI CommunityToolkit: 1.0.1
- OS: Android 10 (SDK 29)
- .NET MAUI: 7.0
```

**Comment by jfversluis** [comment]
> Ridiculously simply to reproduce, faster than a download in this case..

While you might think that, it's often not that simple. Remember, we don't have all the context you have. There are a lot of factors that can influence the behavior of a potential issue.

For instance, in this case, I can't reproduce it. Using this RTSP stream: rtsp://wowzaec2demo.streamlock.net/vod/mp4:BigBuckBunny_115k.mp4 it works perfectly for me.

So it might be something in the way we configure the ExoPlayer or how we load it, but at the very least I'm going to need a RTSP stream that actually shows this behavior.

**Comment by Anioz** [comment]
I updated the comment with a repro link https://1drv.ms/u/s!AqVY9RkzgdHIrRJpCfa_rIgyBW7A?e=8oxTBc but its not showing yet.

Asking to make my usernames passwords and lan camera internet available to all is definatly unrealistic.

Allowing the maui app to feedback more in depth exception info from Exoplayer might be the better way to go, as i stated before i dont mind running experimental builds from you guys to test this and feedback that sensitive information privatly.

**Comment by Anioz** [comment]
I also tested that stream and it works perfectly here locally: rtsp://wowzaec2demo.streamlock.net/vod/mp4:BigBuckBunny_115k.mp4

**Comment by Anioz** [comment]
this one seems very similar: https://github.com/google/ExoPlayer/issues/4532

one user mentions: "In old ExoPlayer i changed ASSERTIONS_ENABLED to false to get rid of crash but now I'm using 2.8.2 and same problem still appeared on Samsung Devices with OS 7.0. ASSERTIONS_ENABLED is public static final in ExoPlayerLibraryInfo and can't modified this time cause I'm using 2.8.2 with gradle implementation."

Is there a possibility for me to tweak and play around with those settings somehow? (to try the proposed work arounds)

**Comment by bijington** [comment]
Yes you can fork this repository and mess around with the settings as much as you like. You can point your app to the build from the fork rather than the NuGet package

**Comment by Anioz** [comment]
> Yes you can fork this repository and mess around with the settings as much as you like. You can point your app to the build from the fork rather than the NuGet package

Thanks to downloading maui 1.0.1 toolkit and modifying the exoplayer constructor as follows:

		var builder = new IExoPlayer.Builder(MauiContext.Context);

		builder.SetMediaSourceFactory(new RtspMediaSource.Factory().SetDebugLoggingEnabled(true));
		Player = builder.Build() ?? throw new NullReferenceException();

		Player.AddListener(this);

Following log item came available:
[RtspClient] RTSP/1.0 401 Unauthorized
[RtspClient] CSeq: 1
[RtspClient] WWW-Authenticate: Digest realm="Hikvision", nonce="a1142a0c619202fe28be22a7c8789a6d", random="G66huST+L3nomzn1NFq/isZmLVyHlh7/6qVGYsmoEHwZniHaisXZIRB/PYXiKg9R", stale="FALSE"
[RtspClient] WWW-Authenticate: Basic realm="Hikvision"
[RtspClient] Date:  Wed, Feb 08 2023 12:34:06 GMT
[RtspClient] DESCRIBE rtsp://1.4.1.23:554/ch1/sub/live RTSP/1.0
[RtspClient] User-Agent: ExoPlayerLib/2.18.2
[RtspClient] CSeq: 2
[RtspClient] Authorization: Digest username="user", realm="Hikvision", nonce="a1142a0c619202fe28be22a7c8789a6d", uri="rtsp://1.4.1.23:554/ch1/sub/live", response="0de7c2e951f5c280dd5bc7af81df8d95"
[RtspClient] 
[RtspClient] RTSP/1.0 200 OK
[RtspClient] CSeq: 2
[RtspClient] Content-Type: application/sdp
[RtspClient] Content-Base: rtsp://10.104.1.63:554/ch1/sub/live/
[RtspClient] Content-Length: 551
[RtspClient] 
[RtspClient] v=0
[RtspClient] o=- 1675859646308821 1675859646308821 IN IP4 10.104.1.63
[RtspClient] s=Media Presentation
[RtspClient] e=NONE
[RtspClient] b=AS:5100
[RtspClient] t=0 0
[RtspClient] a=control:rtsp://1.4.1.23:554/ch1/sub/live/
[RtspClient] m=video 0 RTP/AVP 96
[RtspClient] b=AS:5000
[RtspClient] a=control:rtsp://1.4.1.23:554/ch1/sub/live/trackID=1
[RtspClient] a=rtpmap:96 H264/90000
[RtspClient] a=fmtp:96 profile-level-id=420029; packetization-mode=1; sprop-parameter-sets=
[RtspClient] m=audio 0 RTP/AVP 0
[RtspClient] b=AS:50
[RtspClient] a=control:rtsp://1.4.1.23:554/ch1/sub/live/trackID=2
[RtspClient] a=rtpmap:0 PCMU/8000
[RtspClient] a=Media_header:MEDIAINFO=494D4B48010100000400010010710110401F0000000;
[RtspClient] a=appversion:1.0 

So basically its IGNORING/NOT USING the given password in the pswd string for some reason altough with the other URL's which are the same in several cases (except for the ip and username)

**Comment by Anioz** [comment]
With rtsp authentication DISABLED on the camera that won't work with Exoplayer the following log output is gotten:
[RtspClient] OPTIONS rtsp://1.4.3.23:554/ch1/sub/live RTSP/1.0
[RtspClient] User-Agent: ExoPlayerLib/2.18.2
[RtspClient] CSeq: 0
[RtspClient] 
[RtspClient] RTSP/1.0 200 OK
[RtspClient] CSeq: 0
[RtspClient] Public: OPTIONS, DESCRIBE, PLAY, PAUSE, SETUP, TEARDOWN
[RtspClient] Date:  Wed, Feb 08 2023 13:54:56 GMT
[RtspClient] DESCRIBE rtsp://1.4.3.23:554/ch1/sub/live RTSP/1.0
[RtspClient] User-Agent: ExoPlayerLib/2.18.2
[RtspClient] CSeq: 1
[RtspClient] 
[RtspClient] RTSP/1.0 200 OK
[RtspClient] CSeq: 1
[RtspClient] Content-Type: application/sdp
[RtspClient] Content-Base: rtsp://1.4.3.23:554/ch1/sub/live/
[RtspClient] Content-Length: 551
[RtspClient] 
[RtspClient] v=0
[RtspClient] o=- 1675864496608871 1675864496608871 IN IP4 1.4.3.23
[RtspClient] s=Media Presentation
[RtspClient] e=NONE
[RtspClient] b=AS:5100
[RtspClient] t=0 0
[RtspClient] a=control:rtsp://1.4.3.23:554/ch1/sub/live/
[RtspClient] m=video 0 RTP/AVP 96
[RtspClient] b=AS:5000
[RtspClient] a=control:rtsp://1.4.3.23:554/ch1/sub/live/trackID=1
[RtspClient] a=rtpmap:96 H264/90000
[RtspClient] a=fmtp:96 profile-level-id=420029; packetization-mode=1; sprop-parameter-sets=
[RtspClient] m=audio 0 RTP/AVP 0
[RtspClient] b=AS:50
[RtspClient] a=control:rtsp://1.4.3.23:554/ch1/sub/live/trackID=2
[RtspClient] a=rtpmap:0 PCMU/8000
[RtspClient] a=Media_header:MEDIAINFO=494D4B48010100000400010010710110401F0000000;
[RtspClient] a=appversion:1.0
[SurfaceView] onWindowVisibilityChanged(0) true android.view.SurfaceView{a65b1cd V.E...... ......I. 0,0-0,0} of ViewRootImpl@2d39d02[MainActivity]
[AndroidRuntime] FATAL EXCEPTION: ExoPlayer:Playback
[AndroidRuntime] Process: com.companyname.rtspcrash, PID: 878
[AndroidRuntime] java.lang.IllegalArgumentException
[AndroidRuntime] 	at com.google.android.exoplayer2.util.Assertions.checkArgument(Assertions.java:39)
[AndroidRuntime] 	at com.google.android.exoplayer2.source.rtsp.RtspMediaTrack.processH264FmtpAttribute(RtspMediaTrack.java:384)
[AndroidRuntime] 	at com.google.android.exoplayer2.source.rtsp.RtspMediaTrack.generatePayloadFormat(RtspMediaTrack.java:261)
[AndroidRuntime] 	at com.google.android.exoplayer2.source.rtsp.RtspMediaTrack.<init>(RtspMediaTrack.java:161)
[AndroidRuntime] 	at com.google.android.exoplayer2.source.rtsp.RtspClient.buildTrackList(RtspClient.java:347)
[AndroidRuntime] 	at com.google.android.exoplayer2.source.rtsp.RtspClient.access$1700(RtspClient.java:75)
[AndroidRuntime] 	at com.google.android.exoplayer2.source.rtsp.RtspClient$MessageListener.onDescribeResponseReceived(RtspClient.java:690)
[AndroidRuntime] 	at com.google.android.exoplayer2.source.rtsp.RtspClient$MessageListener.handleRtspResponse(RtspClient.java:598)
[AndroidRuntime] 	at com.google.android.exoplayer2.source.rtsp.RtspClient$MessageListener.handleRtspMessage(RtspClient.java:507)
[AndroidRuntime] 	at com.google.android.exoplayer2.source.rtsp.RtspClient$MessageListener.lambda$onRtspMessageReceived$0$com-google-android-exoplayer2-source-rtsp-RtspClient$MessageListener(RtspClient.java:500)
[AndroidRuntime] 	at com.google.android.exoplayer2.source.rtsp.RtspClient$MessageListener$$ExternalSyntheticLambda0.run(Unknown Source:4)
[AndroidRuntime] 	at android.os.Handler.handleCallback(Handler.java:883)
[AndroidRuntime] 	at android.os.Handler.dispatchMessage(Handler.java:100)
[AndroidRuntime] 	at android.os.Looper.loop(Looper.java:237)
[AndroidRuntime] 	at android.os.HandlerThread.run(HandlerThread.java:67)

**Comment by bijington** [comment]
> Thanks to downloading maui 1.0.1 toolkit and modifying the exoplayer constructor as follows:

	var builder = new IExoPlayer.Builder(MauiContext.Context);

	builder.SetMediaSourceFactory(new RtspMediaSource.Factory().SetDebugLoggingEnabled(true));
	Player = builder.Build() ?? throw new NullReferenceException();

	Player.AddListener(this);

Thanks for this update. It sounds like we could possible add the ability to add in logging information into the documentation pages.

**Comment by Anioz** [comment]
> > Thanks to downloading maui 1.0.1 toolkit and modifying the exoplayer constructor as follows:
> 
> ```
> var builder = new IExoPlayer.Builder(MauiContext.Context);
> 
> builder.SetMediaSourceFactory(new RtspMediaSource.Factory().SetDebugLoggingEnabled(true));
> Player = builder.Build() ?? throw new NullReferenceException();
> 
> Player.AddListener(this);
> ```
> 
> Thanks for this update. It sounds like we could possible add the ability to add in logging information into the documentation pages.

Glad to help. Adding it to the standard documentation will probably be very helpfull for weeding out a lot of the exoplayer type problems

Wish there was a rtsp auth reply msg editor made available in the exoplayer bindings that way we could even solve problems and just feedback the solution in its entirety.

**Comment by dorton21** [comment]
@Anioz Would you be willing to post more information about how you resolved the authentication issues?

**Comment by Anioz** [comment]
> @Anioz Would you be willing to post more information about how you resolved the authentication issues?

I ended up doing away with exoplayer entirely and used a webview instead. For my videoplayer capabilities that i needed for other vids that are not streams i implemented libvlc video player for cross platform maui.
